tools/tcpconnect.py

```diff
@@ -224,7 +231,86 @@
                bpf_get_current_comm(&data6.task, sizeof(data6.task));
                ipv6_events.perf_submit(ctx, &data6, sizeof(data6));"""
                }
-        }
+               }
+
+# This defines an additional BPF program that instruments udp_recvmsg system
+# call to locate DNS response packets on UDP port 53. When these packets are
+# located, the data is copied to user-space where python will parse them with
+# dnslib.
+#
+# uses a percpu array of length 1 to store the dns_data_t off the stack to
+# allow for a maximum DNS packet length of 512 bytes.
+dns_bpf_text = """
+#include <net/inet_sock.h>
+
+#define MAX_PKT 512
+struct dns_data_t {
+    u8  pkt[MAX_PKT];
+};
+
+BPF_PERF_OUTPUT(dns_events);
+
+// store msghdr pointer captured on syscall entry to parse on syscall return
+BPF_HASH(tbl_udp_msg_hdr, u64, struct msghdr *);
+
+// single element per-cpu array to hold the current event off the stack
+BPF_PERCPU_ARRAY(dns_data,struct dns_data_t,1);
+
+int trace_udp_recvmsg(struct pt_regs* ctx)
+{
+    __u64 pid_tgid = bpf_get_current_pid_tgid();
+    struct sock *sk = (struct sock*)PT_REGS_PARM1(ctx);
+    struct inet_sock *is = inet_sk(sk);
+
+    // only grab port 53 packets, 13568 is ntohs(53)
+    if (is->inet_dport == 13568) {
+        struct msghdr *msghdr = (struct msghdr*)PT_REGS_PARM2(ctx);
+        tbl_udp_msg_hdr.update(&pid_tgid, &msghdr);
+    }
+    return 0;
+}
+
+int trace_udp_ret_recvmsg(struct pt_regs* ctx)
+{
+    __u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 zero = 0;
+    struct msghdr **msgpp = tbl_udp_msg_hdr.lookup(&pid_tgid);
```

Comment:
does this always work (matching the PID of the DNS call to the PID of the TCP connect)?

What about systemd-resolved? Are there cases where an application asks a resolver to do name resolution on its behalf? If so, the PID of the DNS != PID of the TCP.

Reply:
Oh, this is for matching entry to ret of udp_recvmsg(), so disregard my comment.